test/e2e/obsctl_test.go

```diff
@@ -0,0 +1,248 @@
+package e2e
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/efficientgo/e2e"
+	"github.com/efficientgo/tools/core/pkg/testutil"
+	"github.com/observatorium/obsctl/pkg/cmd"
+)
+
+const (
+	envName       = "obsctl-test"
+	hydraURL      = "172.17.0.1:4444" // TODO(saswatamcode): Make this macOS-friendly.
+	noOfTenants   = 2                 // Configure number of tenants.
+	defaultTenant = 1                 // Set default tenant to use.
+)
+
+// preTest spins up all services required for metrics:
+// - Receive
+// - Query
+// - Rule
+// - Minio, Rules Objstore, Rule Syncer
+// - Up
+// Hydra is spun up externally via start_hydra.sh, as accessing it via docker network is difficult for obsctl.
+// Follows similar pattern as https://observatorium.io/docs/usage/getting-started.md/.
+// Also registers tenants in hydra.
+func preTest(t *testing.T) *e2e.DockerEnvironment {
+	e, err := e2e.NewDockerEnvironment(envName)
+	testutil.Ok(t, err)
+	t.Cleanup(e.Close)
+
+	err = os.MkdirAll(filepath.Join(e.SharedDir(), "config"), 0750)
+	testutil.Ok(t, err)
+
+	registerHydraUsers(t, noOfTenants) // Only need to register this once.
+
+	createTenantsYAML(t, e, hydraURL, noOfTenants)
+	createRBACYAML(t, e, noOfTenants)
+
+	read, write, rule := startServicesForMetrics(t, e, envName)
+
+	api, err := newObservatoriumAPIService(e, withMetricsEndpoints(read, write), withRulesEndpoint(rule))
+	testutil.Ok(t, err)
+	testutil.Ok(t, e2e.StartAndWaitReady(api))
+	testutil.Ok(t, os.MkdirAll(filepath.Join(e.SharedDir(), "obsctl"), 0750)) // Create config file beforehand.
+
+	createObsctlConfigJson(t, e, hydraURL, "http://"+api.Endpoint("http")+"/", noOfTenants, defaultTenant)
+
+	token := obtainToken(t, hydraURL, defaultTenant)
+
+	up, err := newUpRun(
+		e, "up-metrics-read-write",
+		"http://"+api.InternalEndpoint("http")+"/api/metrics/v1/test-oidc-"+fmt.Sprint(defaultTenant)+"/api/v1/query",
+		"http://"+api.InternalEndpoint("http")+"/api/metrics/v1/test-oidc-"+fmt.Sprint(defaultTenant)+"/api/v1/receive",
+		withToken(token),
+		withRunParameters(&runParams{period: "500ms", threshold: "1", latency: "10s", duration: "0"}),
+	)
+
+	createRulesYAML(t, e)
+
+	testutil.Ok(t, e2e.StartAndWaitReady(up))
+	testutil.Ok(t, err)
+
+	time.Sleep(30 * time.Second) // Wait a bit for up to get some metrics in.
+
+	return e
+}
+
+func TestObsctlMetricsCommands(t *testing.T) {
+	e := preTest(t)
+
+	t.Run("get labels for a tenant", func(t *testing.T) {
+		b := bytes.NewBufferString("")
+
+		contextCmd := cmd.NewObsctlCmd(context.TODO(), filepath.Join(e.SharedDir(), "obsctl", "config.json"))
```

Comment:
Why just not `context.Background` for all of the test cases?

Reply:
Afaik, context.TODO() and context.Background() [return the same thing](https://cs.opensource.google/go/go/+/refs/tags/go1.18.1:src/context/context.go;l=208). But yeah this should be background as we are deliberately passing an empty context. Let me change!

Reply:
Done in https://github.com/observatorium/obsctl/pull/12/commits/22d7e63bd5e626f25203c365b57258e05e4123ad

Reply:
Hmm, test fails but passes locally. Also response is the same, so maybe just a flake somewhere?

Reply:
They do return same thing but signify different things - https://pkg.go.dev/context#TODO.

Yeah, looks like one of the tests is flaky already :thinking:

Reply:
Yeah, but I tried running multiple times and wasn't able to replicate. Maybe let's re-run here to confirm?

Reply:
Started a re-run :+1: I'm just wondering if it isn't one of those flakes that will slowly start appearing more often on our CI due to e.g. CI machines being under heavy load.